File: app/src/main/java/com/balch/auctionbrowser/auction/model/AuctionDataTypeAdapter.java
package com.balch.auctionbrowser.auction.model;

import com.balch.android.app.framework.types.ISO8601DateTime;
import com.balch.android.app.framework.types.Money;
import com.balch.auctionbrowser.AuctionData;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;

import java.lang.reflect.Type;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class AuctionDataTypeAdapter implements JsonDeserializer<AuctionData> {
    @Override
    public AuctionData deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
            throws JsonParseException {

        JsonObject items = json.getAsJsonObject()
                .getAsJsonArray("findItemsByKeywordsResponse").get(0)
                .getAsJsonObject();
        boolean success = validateResponse(items);

        AuctionData auctionData = new AuctionData();
        if (success) {
            auctionData.setTotalPages(getTotalPages(items));
            auctionData.setAuctions(parseAuctions(items));
        } else {
        }

        return auctionData;
    }

    private boolean validateResponse(JsonObject json) {
        return ((json != null) &&
                json.getAsJsonArray("ack").get(0).getAsString().equals("Success"));
    }

    private int getTotalPages(JsonObject json) {
        JsonPrimitive totalPages = json
                .getAsJsonArray("paginationOutput").get(0).getAsJsonObject()
                .getAsJsonArray("totalPages").get(0).getAsJsonPrimitive();
        return totalPages.getAsInt();

    }

    private List<Auction> parseAuctions(JsonObject json) throws JsonParseException {
        List<Auction> auctions = new ArrayList<>();
        JsonArray items = json
                .getAsJsonArray("searchResult").get(0).getAsJsonObject()
                .getAsJsonArray("item");
        for(JsonElement element : items){
            auctions.add(parseAuction(element.getAsJsonObject()));
        }
        return auctions;
    }

    private Auction parseAuction(JsonObject item) throws JsonParseException {

        JsonArray jsonViewItemURL = item.getAsJsonArray("viewItemURL");
        JsonArray jsonGalleryURL = item.getAsJsonArray("galleryURL");
        JsonArray jsonLocaton = item.getAsJsonArray("location");
        JsonObject sellingStatus = item.getAsJsonArray("sellingStatus").get(0).getAsJsonObject();
        double convertedCurrentPrice = sellingStatus
                .getAsJsonArray("convertedCurrentPrice").get(0).getAsJsonObject()
                .getAsJsonPrimitive("__value__").getAsDouble();

        Money shippingCost = new Money();
        JsonArray shippingInfo = item.getAsJsonArray("shippingInfo");
        if (shippingInfo != null) {
            JsonArray shippingServiceCost = shippingInfo.get(0).getAsJsonObject()
                    .getAsJsonArray("shippingServiceCost");
            if (shippingServiceCost != null) {
                double cost = shippingServiceCost.get(0).getAsJsonObject()
                        .getAsJsonPrimitive("__value__").getAsDouble();
                shippingCost = new Money(cost);
            }
        }

        Boolean isAuction = false;
        Boolean isButItNow = false;
        Date startTime = new Date();
        Date endTime = new Date();
        JsonArray listingInfoWrapper = item.getAsJsonArray("listingInfo");
        if (listingInfoWrapper != null) {
            JsonObject listingInfo = listingInfoWrapper.get(0).getAsJsonObject();
            String listingType = listingInfo.getAsJsonArray("listingType").get(0).getAsString();

            if ("Auction".equals(listingType)) {
                isAuction = true;
                isButItNow = listingInfo.getAsJsonArray("buyItNowAvailable").get(0).getAsBoolean();
            } else {
                isAuction = false;
                isButItNow = true;
            }

            try {
                startTime = ISO8601DateTime.toDate(listingInfo.getAsJsonArray("startTime").get(0).getAsString());
                endTime = ISO8601DateTime.toDate(listingInfo.getAsJsonArray("endTime").get(0).getAsString());
            } catch (ParseException ex) {
                throw new JsonParseException(ex);
            }
        }

        Auction auction = new Auction(
            item.getAsJsonArray("itemId").get(0).getAsLong(),
            item.getAsJsonArray("title").get(0).getAsString(),
            (jsonViewItemURL != null) ? jsonViewItemURL.get(0).getAsString() : "",
            (jsonGalleryURL != null) ? jsonGalleryURL.get(0).getAsString() : "",
            (jsonLocaton != null) ? jsonLocaton.get(0).getAsString() : "",
            shippingCost,
            new Money(convertedCurrentPrice),
            "",
            startTime,
            endTime,
            isAuction,
            isButItNow);

        return auction;
    }

}

File: app/src/main/java/com/balch/auctionbrowser/auction/model/AuctionDataTypeAdapter.kt
/*
 * Author: Balch
 *
 * This file is part of AuctionBrowser.
 *
 * AuctionBrowser is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * AuctionBrowser is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MockTrade.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2017
 *
 */

package com.balch.auctionbrowser.auction.model

import com.balch.auctionbrowser.types.Money
import com.google.gson.*
import org.joda.time.DateTime
import java.lang.reflect.Type
import java.text.ParseException
import java.util.*

/**
 * Retrofit parser for the EBayApi.
 */
class AuctionDataTypeAdapter : JsonDeserializer<AuctionData> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): AuctionData {

        val items = json.asJsonObject
                .getAsJsonArray("findItemsByKeywordsResponse").get(0)
                .asJsonObject
        val success = validateResponse(items)

        return if (success)
            AuctionData().apply {
                totalPages = getTotalPages(items)
                auctions = parseAuctions(items)
            }
        else
            AuctionData().apply {
                hasError = true
            }
    }

    private fun validateResponse(json: JsonObject?): Boolean {
        return json != null && json.getAsJsonArray("ack").get(0).asString == "Success"
    }

    private fun getTotalPages(json: JsonObject): Int {
        val totalPages = json
                .getAsJsonArray("paginationOutput").get(0).asJsonObject
                .getAsJsonArray("totalPages").get(0).asJsonPrimitive
        return totalPages.asInt
    }

    @Throws(JsonParseException::class)
    private fun parseAuctions(json: JsonObject): List<Auction> {
        val items = json
                .getAsJsonArray("searchResult").get(0).asJsonObject
                .getAsJsonArray("item")
        val auctions = items.mapTo(ArrayList<Auction>()) { parseAuction(it.asJsonObject) }
        return auctions
    }

    @Throws(JsonParseException::class)
    private fun parseAuction(item: JsonObject): Auction {

        val jsonViewItemURL = item.getAsJsonArray("viewItemURL")
        val jsonGalleryURL = item.getAsJsonArray("galleryURL")
        val jsonLocaton = item.getAsJsonArray("location")
        val sellingStatus = item.getAsJsonArray("sellingStatus").get(0).asJsonObject
        val convertedCurrentPrice = sellingStatus
                .getAsJsonArray("convertedCurrentPrice").get(0).asJsonObject
                .getAsJsonPrimitive("__value__").asDouble

        var shippingCost = Money()
        val shippingInfo = item.getAsJsonArray("shippingInfo")
        if (shippingInfo != null) {
            val shippingServiceCost = shippingInfo.get(0).asJsonObject
                    .getAsJsonArray("shippingServiceCost")
            if (shippingServiceCost != null) {
                val cost = shippingServiceCost.get(0).asJsonObject
                        .getAsJsonPrimitive("__value__").asDouble
                shippingCost = Money(cost)
            }
        }

        var isAuction: Boolean = false
        var isButItNow: Boolean = false
        var startTime = DateTime()
        var endTime = DateTime()
        val listingInfoWrapper = item.getAsJsonArray("listingInfo")
        if (listingInfoWrapper != null) {
            val listingInfo = listingInfoWrapper.get(0).asJsonObject
            val listingType = listingInfo.getAsJsonArray("listingType").get(0).asString

            if ("Auction" == listingType) {
                isAuction = true
                isButItNow = listingInfo.getAsJsonArray("buyItNowAvailable").get(0).asBoolean
            } else {
                isAuction = false
                isButItNow = true
            }

            try {
                startTime = DateTime.parse(listingInfo.getAsJsonArray("startTime").get(0).asString)
                endTime = DateTime.parse(listingInfo.getAsJsonArray("endTime").get(0).asString)
            } catch (ex: ParseException) {
                throw JsonParseException(ex)
            }

        }

        val auction = Auction(
                item.getAsJsonArray("itemId").get(0).asLong,
                item.getAsJsonArray("title").get(0).asString,
                if (jsonGalleryURL != null) jsonGalleryURL.get(0).asString else "",
                if (jsonViewItemURL != null) jsonViewItemURL.get(0).asString else "",
                if (jsonLocaton != null) jsonLocaton.get(0).asString else "",
                shippingCost,
                Money(convertedCurrentPrice),
                "",
                startTime,
                endTime,
                isAuction,
                isButItNow)

        return auction
    }

}
